src/cas/utils/validation_utils.py

```diff
@@ -72,3 +84,263 @@ def validate_labelset_markers(annotation: Dict[str, Any], marker_list: List[str]
+
+    Parameters:
+        cas: The CAS JSON object.
+        obs: The AnnData obs DataFrame.
+
+    Returns:
+        True if all labelsets from CAS exist as columns in obs, otherwise False.
+    """
+    labelset_list = [labelset[LABELSET_NAME] for labelset in cas[LABELSETS]]
+    obs_columns = set(obs.columns)
+
+    # Find missing labelsets
+    missing_labelsets = set(labelset_list) - obs_columns
+
+    if missing_labelsets:
+        logging.warning(f"Missing labelsets in obs: {missing_labelsets}")
+        return False
+    else:
+        logging.info("All labelsets exist in obs.")
+        return True
+
+
+def validate_labelset_values(
+    cas: Dict[str, Any], obs: Union[pd.DataFrame, CapAnnDataDF]
+) -> bool:
+    """
+    Validate that all labelset members from CAS annotations exist in the corresponding obs labelset columns.
+
+    Logs warnings for any missing labelset members.
+
+    Parameters:
+        cas (Dict[str, Any]): The CAS JSON object.
+        obs (pd.DataFrame): The AnnData obs DataFrame.
+
+    Returns:
+        True if all labelset members from CAS exist in obs, otherwise False.
+    """
+    labelset_members = {}
+
+    # Collect all labelset members from CAS annotations
+    for annotation in cas[ANNOTATIONS]:
+        labelset_name = annotation[LABELSET]
+        cell_label = annotation[CELL_LABEL]
+
+        if labelset_name not in labelset_members:
+            labelset_members[labelset_name] = set()
+
+        labelset_members[labelset_name].add(cell_label)
+
+    # Check if all labelset members exist in obs
+    missing_values = {}
+
+    for labelset_name, members in labelset_members.items():
+        if labelset_name in obs.columns:
+            obs_values = set(obs[labelset_name].dropna().unique())
+            missing_members = members - obs_values
+
+            if missing_members:
+                missing_values[labelset_name] = missing_members
+
+    # Log warnings for missing values
+    if missing_values:
+        for labelset, missing in missing_values.items():
+            logging.warning(f"Missing labelset members in obs['{labelset}']: {missing}")
+        return False
+    else:
+        logging.info("All labelset members exist in the corresponding obs columns.")
+        return True
+
+
+def check_parent_child_consistency(
+    cas: Dict[str, Any], obs: Union[pd.DataFrame, CapAnnDataDF]
+) -> None:
+    """
+    Checks if the inferred hierarchy from cell labels in obs matches the expected hierarchy from
+    CAS rank data.
+
+    Parameters:
+        cas (Dict[str, Any]): The CAS JSON object containing labelset ranks.
+        obs (pd.DataFrame): The AnnData obs DataFrame.
+
+    Returns:
+        True if all inferred parent-child relationships from obs match those from cas,
+              otherwise False.
+    """
+    # Extract ranks from CAS (store expected rank per cell label)
+    cas_ranks = {entry["name"]: entry["rank"] for entry in cas[LABELSETS]}
```

Comment:
need to handle labelsets without a rank

Reply:
How? Where do they fit into the hierarchy?

Reply:
They are not hierarchical but `entry["rank"]` will throw an exception:
```
{entry["name"]: entry["rank"] for entry in cas[LABELSETS] if "rank" in entry}
```